lib/timers.js

```diff
@@ -226,122 +284,105 @@ function listOnTimeout(handle, now) {
-    delete refedLists[msecs];
-  }
-
-  // Do not close the underlying handle if its ownership has changed
-  // (e.g it was unrefed in its callback).
-  if (!handle.owner)
-    handle.close();
 
-  return true;
+  // The current list may have been removed and recreated since the reference
+  // to `list` was created. Make sure they're the same instance of the list
+  // before destroying.
+  if (list === lists[msecs]) {
+    delete lists[msecs];
+    queue.shift();
+  }
 }
 
 
 // An optimization so that the try/finally only de-optimizes (since at least v8
 // 4.7) what is in this smaller function.
 function tryOnTimeout(timer, start) {
-  timer._called = true;
-  const timerAsyncId = (typeof timer[async_id_symbol] === 'number') ?
-    timer[async_id_symbol] : null;
-  var threw = true;
-  if (timerAsyncId !== null)
-    emitBefore(timerAsyncId, timer[trigger_async_id_symbol]);
   if (start === undefined && timer._repeat)
     start = TimerWrap.now();
   try {
     ontimeout(timer);
-    threw = false;
   } finally {
-    if (timerAsyncId !== null) {
-      if (!threw)
-        emitAfter(timerAsyncId);
-      if (timer._repeat) {
-        rearm(timer, start);
-      } else if (destroyHooksExist() && !timer._destroyed) {
-        emitDestroy(timerAsyncId);
+    if (timer._repeat) {
+      rearm(timer, start);
+    } else {
+      if (timer[kRefed])
+        refCount--;
+      timer[kRefed] = null;
+
+      if (destroyHooksExist() && !timer._destroyed) {
+        emitDestroy(timer[async_id_symbol]);
         timer._destroyed = true;
       }
     }
   }
 }
 
 
-// A convenience function for re-using TimerWrap handles more easily.
-//
-// This mostly exists to fix https://github.com/nodejs/node/issues/1264.
-// Handles in libuv take at least one `uv_run` to be registered as unreferenced.
-// Re-using an existing handle allows us to skip that, so that a second `uv_run`
-// will return no active handles, even when running `setTimeout(fn).unref()`.
-function reuse(item) {
-  L.remove(item);
-
-  const list = refedLists[item._idleTimeout];
-  // if empty - reuse the watcher
-  if (list !== undefined && L.isEmpty(list)) {
-    debug('reuse hit');
-    list._timer.stop();
-    delete refedLists[item._idleTimeout];
-    return list._timer;
-  }
-
-  return null;
-}
-
-
 // Remove a timer. Cancels the timeout and resets the relevant timer properties.
 function unenroll(item) {
   // Fewer checks may be possible, but these cover everything.
   if (destroyHooksExist() &&
-      typeof item[async_id_symbol] === 'number' &&
+      item[async_id_symbol] !== undefined &&
       !item._destroyed) {
     emitDestroy(item[async_id_symbol]);
     item._destroyed = true;
   }
 
-  const handle = reuse(item);
-  if (handle !== null) {
-    debug('unenroll: list empty');
-    handle.close();
+  L.remove(item);
+
+  // We only delete refed lists because unrefed ones are incredibly likely
+  // to come from http and be recreated shortly after.
```

Comment:
Isn't this a memory leak?

Also, in most situations where pref matters (e.g. under load) won't enough connections be present that timeouts will usually overlap, causing the list to stay alive anyways?

Reply:
This just keeps existing behaviour. See `reuse`.

Reply:
Right. Ok, I recall a thing about that - we keep the unrefed list "alive" and let libuv clean it up, right?

We might want to test that it is indeed cleaned up when insert is run when that list should be hit (it should be cleaned up at that point).

Reply:
That's correct, yeah.

> We might want to test that it is indeed cleaned up when insert is run when that list should be hit (it should be cleaned up at that point).

It does work this way. I suppose we could create a test but it would probably necessitate exporting more stuff from internal timers.